pkg/restore/conflict.go

```diff
@@ -0,0 +1,192 @@
+			// Copy all secrets *except* default-token
+			secrets = append(secrets[:i], secrets[i+1:]...)
+			break
+		}
+	}
+	desiredSecrets, _, err := unstructured.NestedSlice(desired.UnstructuredContent(), "secrets")
+	if err != nil {
+		return nil, err
+	}
+	desired.Object["secrets"] = collections.MergeSlices(desiredSecrets, secrets, mapsAreEqual)
+
+	backupImageSecrets, _, err := unstructured.NestedSlice(fromBackup.UnstructuredContent(), "imagePullSecrets")
+	if err != nil {
+		return nil, err
+	}
+	desiredImageSecrets, _, err := unstructured.NestedSlice(desired.UnstructuredContent(), "imagePullSecrets")
+	if err != nil {
+		return nil, err
+	}
+	desired.Object["imagePullSecrets"] = collections.MergeSlices(desiredImageSecrets, backupImageSecrets, mapsAreEqual)
+
+	desiredLabels := desired.GetLabels()
+	backupLabels := fromBackup.GetLabels()
+	collections.MergeMaps(desiredLabels, backupLabels)
+	desired.SetLabels(desiredLabels)
+
+	desiredAnnotations := desired.GetAnnotations()
+	backupAnnotations := fromBackup.GetAnnotations()
+	collections.MergeMaps(desiredAnnotations, backupAnnotations)
+	desired.SetAnnotations(desiredAnnotations)
+
+	return desired, nil
+}
+func mergeServiceAccounts(fromCluster, fromBackup *unstructured.Unstructured) (*unstructured.Unstructured, error) {
+	// merge any labels, all secrets except the default token, and image pull secrets from the backup
+	desired := new(corev1api.ServiceAccount)
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(fromCluster.UnstructuredContent(), desired); err != nil {
+		return nil, errors.Wrap(err, "unable to convert from unstructured to serviceaccount")
+
+	}
+
+	backupSA := new(corev1api.ServiceAccount)
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(fromBackup.UnstructuredContent(), backupSA); err != nil {
+		return nil, errors.Wrap(err, "unable to convert from unstructured to serviceaccount")
+
+	}
+
+	for i := len(backupSA.Secrets) - 1; i >= 0; i-- {
+		secret := backupSA.Secrets[i]
+		if strings.HasPrefix(secret.Name, "default-token-") {
+			// Copy all secrets *except* default-token
+			backupSA.Secrets = append(backupSA.Secrets[:i], backupSA.Secrets[i+1:]...)
+			break
+		}
+	}
+
+	desired.Secrets = mergeObjRefSlices(desired.Secrets, backupSA.Secrets)
+
+	desired.ImagePullSecrets = mergeLocalObjRefSlices(desired.ImagePullSecrets, backupSA.ImagePullSecrets)
+
+	collections.MergeMaps(desired.Labels, backupSA.Labels)
+
+	collections.MergeMaps(desired.Annotations, backupSA.Annotations)
+
+	desiredUnstructured, err := runtime.DefaultUnstructuredConverter.ToUnstructured(desired)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to convert service account to unstructured")
+	}
+
+	return &unstructured.Unstructured{Object: desiredUnstructured}, nil
+}
+
+func mergeObjRefSlices(first, second []corev1api.ObjectReference) []corev1api.ObjectReference {
+	merged := append(make([]corev1api.ObjectReference, 0), first...)
```

Comment:
Changing to `merged := first` in all the merge functions dropped allocations by 1-2 per op, saved on some bytes per op, and didn't noticeably impact execution time.

Reply:
I'm guessing you are trying to avoid modifying `first`. It appears that doing `merged := first` and then appending to `merged` leaves `first` unchanged. You should probably do an independent test 😄